DPL: full serialization of DataDescriptorMatcher in the worklfow dump

matthiasrichter opened the pull request:
working version, cleanup to be done and unit test to be included in PR

The format for serialized matchers is something like:
```
"matcher": {
    "operation": "and",
    "left": {
        "origin": "TST"
    },
    "matcher": {
        "operation": "and",
        "left": {
            "description": "SAMPLE"
        },
        "matcher": {
            "operation": "just",
            "left": {
                "starttime": "$0"
            }
        }
    }
```
Fully qualified input specs which are represented by `ConcreteDataMatcher` are still saved in the simplified format:
```
"origin": "TST",
"description": "SAMPLE",
"subspec": 1,
```
We can add a helper to boil the matcher down to the most possible qualified representation, returning `ConcreteDataMatcher`, `ConreteDataTypeMatcher`, or `OriginValueMatcher`. And with this store the relevant keys in simplified format. But the existing helpers did not yet support this.

matthiasrichter committed:
DPL WorkflowSerializationHelpers: import serialized DataDescriptorMatcher

matthiasrichter committed:
Runtime exception if there is no string handling in the current state

matthiasrichter committed:
Implementing the serialization of DataDescriptorMatcher objects

ktf reviewed (commented):
Looks like the issue is genuine?

matthiasrichter committed:
fixup! Implementing the serialization of DataDescriptorMatcher objects

When activating the simplyfied json format agaion for fully qualified matchers,
the handling of potentially empty objects by inserting placeholder matchers
can not correctly handle this. But meanwhile, empty objects are excluded by
also serializing the StartTimeValueMatcher and we can remove this code.

matthiasrichter commented:
yes, just before pushing, I have reactivated the old json format for the fully qualified descriptors without testing again, so this was a real problem.

matthiasrichter commented:
@ktf does any other component rely on the specific json format? Or is this completely DPL internal?

matthiasrichter commented:
Cant really think that the error in the CS8 CI is related to the PR, Looks like that FairMQ sends a parts object without any messages, can this happen?
```
[24852:dest]: o2-test-framework-CallbackService: /sw/SOURCES/O2/8733-slc8_x86-64/0/Framework/Core/src/DataProcessingDevice.cxx:1113: static void o2::framework::DataProcessingDevice::handleData(o2::framework::DataProcessorContext&, o2::framework::InputChannelInfo&): Assertion `info.parts.Size() > 0' failed.
```
